File: src/pool/runtime/contracts.py
from __future__ import annotations;

from typing import Any, Literal, TypedDict;

GAME_TYPE_POOL = "pool";
MODE_PVP = "pvp";
MODE_PASS_AND_PLAY = "pnp";
VALID_MODES = frozenset({MODE_PVP, MODE_PASS_AND_PLAY});
STATUS_ACTIVE = "active";
STATUS_FINISHED = "finished";
POOL_CLOCK_INITIAL_SECONDS = 900;
INITIAL_CLOCK_MS = POOL_CLOCK_INITIAL_SECONDS * 1000;
MAX_SHOT_VELOCITY = 4000;
SHOT_VEL_EPSILON = 1.0;

Mode = Literal["pvp", "pnp"];
Status = Literal["active", "finished", "draw"];
Color = Literal["w", "b"];
Seat = Literal["player1", "player2"];

class BallPosition(TypedDict):
    """Persisted position for one numbered Pool ball.""";

    n: int;
    x: float;
    y: float;

class AimMetadata(TypedDict, total=False):
    """Optional frontend aim metadata stored with a shot trajectory.""";

    angle: float;
    power: float;
    cue_x: float;
    cue_y: float;

class Trajectory(TypedDict, total=False):
    """Physics trajectory JSON; frame details remain engine-owned.""";

    x_vel: float;
    y_vel: float;
    aim: AimMetadata;
    frames: list[dict[str, Any]];

class PoolTable(TypedDict, total=False):
    """Serialized Pool table exposed to clients and persisted without SVG/photos.""";

    physics_game_id: str;
    game_name: str;
    mode: Mode;
    status: Status;
    game_over: bool;
    game_started: bool;
    winner: str | None;
    winner_player_id: str | None;
    winner_message: str;
    p1_name: str;
    p1_playing: str;
    p1_score: int;
    p2_name: str;
    p2_playing: str;
    p2_score: int;
    current_player: str;
    current_player_id: str;
    player1_id: str;
    player2_id: str;
    p1_photo_url: str | None;
    p2_photo_url: str | None;
    ball_in_hand_for_player_id: str | None;
    ball_in_hand: bool;
    svg: str;
    balls: list[BallPosition];

class SeatFlags(TypedDict):
    """Per-viewer action permissions derived from table state and seat.""";

    can_place_cue: bool;
    can_fire_shot: bool;

class PoolResult(TypedDict, total=False):
    """Terminal result embedded in authoritative Pool state and SSE events.""";

    status: Status;
    winner: str | None;
    reason: str | None;

class PoolState(TypedDict, total=False):
    """Authoritative HTTP/SSE Pool state for one game.""";

    ok: bool;
    game_id: str;
    mode: Mode;
    status: Status;
    result: PoolResult;
    table: PoolTable;
    you_seat: Seat | None;
    stream_seq: int;
    clock_a_ms: int;
    clock_b_ms: int;
    clock_active_color: Color;
    clock_anchor_iso: str;
    can_place_cue: bool;
    can_fire_shot: bool;
    _internal_realtime_next_seq: int;

class PoolPersistPayload(TypedDict, total=False):
    """Async persistence queue item for one Pool table transition.""";

    game_id: str;
    table: PoolTable;
    trajectory: Trajectory | None;
    label: str;
    played_by_id: str | None;
    next_player_id: str | None;
    time_a_ms: int;
    time_b_ms: int;
    mode: Mode;
    owner_user_id: str;
    winner_player_id: str | None;

class ShotAccepted(TypedDict, total=False):
    """HTTP shot/cue response returned to the acting Pool client.""";

    ok: bool;
    valid: bool;
    game_id: str;
    trajectory: Trajectory | None;
    table: PoolTable;
    stream_seq: int;
    clock_a_ms: int;
    clock_b_ms: int;
    clock_active_color: Color;
    clock_anchor_iso: str;
    can_place_cue: bool;
    can_fire_shot: bool;

class PoolGameOverResponse(PoolState, total=False):
    """HTTP surrender/draw response with the final state repeated for clients.""";

    type: Literal["game_over"];
    game_over: PoolState;

class PoolEvent(TypedDict, total=False):
    """Seat-neutral Pool SSE event.""";

    type: Literal["shot_accepted", "game_over", "chat_message"];
    game_id: str;
    next_state: PoolState;
    trajectory: Trajectory | None;
    last_shot_seq: int;
    result: PoolResult;
    message: dict[str, Any];

class ReplayEntry(TypedDict, total=False):
    """One Pool replay position with optional preceding trajectory.""";

    index: int;
    step: int;
    label: str;
    snapshot: PoolTable;
    trajectory: Trajectory | None;

class ReplayTimeline(TypedDict, total=False):
    """Pool replay response assembled from persisted move rows.""";

    ok: bool;
    game_id: str;
    entries: list[ReplayEntry];
    live_index: int;
    trajectories: list[dict[str, Any]];

class SubscribePayload(TypedDict):
    """Long-poll/SSE batch returned by the Pool session coordinator.""";

    game_id: str;
    events: list[dict[str, Any]];
    last_seq: int;

File: src/platform/frontend/src/client/pool/game/contracts.ts
export const MODE_PVP = "pvp" as const;
export const MODE_PASS_AND_PLAY = "pnp" as const;
export type PoolMode = typeof MODE_PVP | typeof MODE_PASS_AND_PLAY;
export type PoolStatus = "active" | "finished" | "draw";
export type PoolSeat = "player1" | "player2";
export type ClockColor = "w" | "b";

export interface BallPosition {
  n: number;
  x: number;
  y: number;
}

export interface ShotAim {
  angle?: number;
  power?: number;
  cue_x?: number;
  cue_y?: number;
  [key: string]: unknown;
}

export interface TrajectorySample {
  t?: number;
  balls?: Record<string, { x: number; y: number }>;
  [key: string]: unknown;
}

export interface PoolTrajectory {
  x_vel?: number;
  y_vel?: number;
  duration?: number;
  aim?: ShotAim;
  samples?: TrajectorySample[];
  [key: string]: unknown;
}

export interface PoolTable {
  game_name?: string;
  svg?: string;
  mode?: PoolMode;
  status?: PoolStatus;
  game_over?: boolean;
  winner?: string | null;
  winner_player_id?: string | null;
  winner_message?: string;
  current_player?: string;
  current_player_id?: string;
  player1_id?: string;
  player2_id?: string;
  p1_name?: string;
  p2_name?: string;
  p1_photo_url?: string | null;
  p2_photo_url?: string | null;
  p1_playing?: string;
  p2_playing?: string;
  p1_score?: number;
  p2_score?: number;
  ball_in_hand?: boolean;
  ball_in_hand_for_player_id?: string | null;
  balls?: BallPosition[];
  [key: string]: unknown;
}

export interface PoolResult {
  status?: PoolStatus;
  winner?: string | null;
  reason?: string | null;
}

export interface PoolState {
  game_id?: string;
  mode?: PoolMode;
  status?: PoolStatus;
  table?: PoolTable;
  result?: PoolResult;
  you_seat?: PoolSeat | null;
  stream_seq?: number;
  clock_a_ms?: number;
  clock_b_ms?: number;
  clock_active_color?: ClockColor;
  clock_anchor_iso?: string;
  can_place_cue?: boolean;
  can_fire_shot?: boolean;
  [key: string]: unknown;
}

export interface PoolActionResponse extends PoolState {
  ok?: boolean;
  valid?: boolean;
  trajectory?: PoolTrajectory | null;
}

export interface PoolTerminalResponse extends Partial<PoolState> {
  type?: string;
  next_state?: Partial<PoolState>;
  game_over?: Partial<PoolState>;
}

export interface PoolReplay {
  game_id: string;
  entries: PoolReplayEntry[];
  live_index: number;
}

export interface PoolReplayEntry {
  index?: number;
  step?: number;
  ply?: number;
  label?: string;
  snapshot?: PoolTable;
  trajectory?: PoolTrajectory | null;
}

export const ACTION = {
  HYDRATE: "HYDRATE",
  SHOT_RESULT: "SHOT_RESULT",
  TERMINAL: "TERMINAL",
  PROFILE_REFRESH: "PROFILE_REFRESH",
} as const;

export type PoolActionType = typeof ACTION[keyof typeof ACTION];

export interface PoolStoreAction {
  type?: PoolActionType;
  payload?: Partial<PoolState>;
}
